### PR TITLE
HLRest: Add get index templates API

### PR DESCRIPTION
**Pull request by dnhatn**
Relates #27205

### CODE REVIEW

File: server/src/test/java/org/elasticsearch/action/admin/indices/template/get/GetIndexTemplatesResponseTests.java [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+    @Override
+    protected GetIndexTemplatesResponse createTestInstance() {
+        List<IndexTemplateMetaData> templates = new ArrayList<>();
+        int numTemplates = between(1, 10);
+        for (int t = 0; t < numTemplates; t++) {
+            IndexTemplateMetaData.Builder templateBuilder = IndexTemplateMetaData.builder("template-" + t);
+            templateBuilder.patterns(IntStream.range(0, between(1, 5)).mapToObj(i -> "pattern-" + i).collect(Collectors.toList()));
+            int numAlias = between(0, 5);
+            for (int i = 0; i < numAlias; i++) {
+                AliasMetaData.Builder alias = AliasMetaData.builder(randomAlphaOfLengthBetween(1, 10));
```
Comment (cbuescher) [inline server/src/test/java/org/elasticsearch/action/admin/indices/template/get/GetIndexTemplatesResponseTests.java]: Maybe its enough to randomize the name here, the full alias parsing is tested more thouroughly somewhere else I think.

**Review by cbuescher — CHANGES_REQUESTED** [review]
Hi @dnhatn, thanks for the PR. I took a quick look and didn't find anything major to change, only left a couple of minor comments which I'm sure are sorted out quickly.

**Comment by dnhatn** [comment]
@cbuescher Thanks for taking a look. I've addressed all your comments.

**Review by cbuescher — APPROVED** [review]
Thanks @dnhatn, LGTM.

File: client/rest-high-level/src/main/java/org/elasticsearch/client/IndicesClient.java [hunk 2]
```diff
@@ -1059,4 +1061,28 @@ public void putTemplateAsync(PutIndexTemplateRequest putIndexTemplateRequest, Re
...
             PutIndexTemplateResponse::fromXContent, listener, emptySet());
     }
+
+    /**
+     * Gets index templates using the Index Templates API
+     * <p>
+     * See <a href="https://www.elastic.co/guide/en/elasticsearch/reference/current/indices-templates.html"> Index Templates API
+     * on elastic.co</a>
+     */
+    public GetIndexTemplatesResponse getTemplates(GetIndexTemplatesRequest getIndexTemplatesRequest,
```
Comment (javanna) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/IndicesClient.java]: nit: can we call it getTemplate? That's what our SPEC have, without the final 's'

File: client/rest-high-level/src/main/java/org/elasticsearch/client/IndicesClient.java [hunk 3]
```diff
@@ -1059,4 +1061,28 @@ public void putTemplateAsync(PutIndexTemplateRequest putIndexTemplateRequest, Re
...
+            options, GetIndexTemplatesResponse::fromXContent, emptySet());
+    }
+
+    /**
+     * Asynchronously gets index templates using the Index Templates API
+     * <p>
+     * See <a href="https://www.elastic.co/guide/en/elasticsearch/reference/current/indices-templates.html"> Index Templates API
+     * on elastic.co</a>
+     */
+    public void getTemplatesAsync(GetIndexTemplatesRequest getIndexTemplatesRequest, RequestOptions options,
```
Comment (javanna) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/IndicesClient.java]: can you improve the javadocs like we have it in the other methods?

File: server/src/test/java/org/elasticsearch/action/admin/indices/template/get/GetIndexTemplatesResponseTests.java [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+                }
+            }
+            templates.add(templateBuilder.build());
+        }
+        return new GetIndexTemplatesResponse(templates);
+    }
+
+    @Override
+    protected boolean supportsUnknownFields() {
+        return false;
```
Comment (javanna) [inline server/src/test/java/org/elasticsearch/action/admin/indices/template/get/GetIndexTemplatesResponseTests.java]: I wonder if it's worth to return true here and exclude some specific paths. for instance at the top-level we can't inject anything arbitrary, within a template it should be possible, within mappings it is not... etc. this is important to verify that we our parsing is forward compatible.

**Review by javanna — APPROVED** [review]
left a couple of comments, LGTM otherwise

File: server/src/test/java/org/elasticsearch/action/admin/indices/template/get/GetIndexTemplatesResponseTests.java (hunk 4 above)
Comment (cbuescher) [reply server/src/test/java/org/elasticsearch/action/admin/indices/template/get/GetIndexTemplatesResponseTests.java]: For what its worth, should those kind of test be part of the IndexTemplateMetaDataTests instead? The response here just renders a list of IndexTemplateMetaData instances with arbitrary names, so in this level are trying to parse every field name as a IndexTemplateMetaData. The parser in this class will indead fail on anything that it encounters thats not parseable as IndexTemplateMetaData, and at the moment it should I think.

Comment (dnhatn) [reply server/src/test/java/org/elasticsearch/action/admin/indices/template/get/GetIndexTemplatesResponseTests.java]: IndexTemplateMetaDataTests already covers these situations.

File: client/rest-high-level/src/main/java/org/elasticsearch/client/IndicesClient.java (hunk 2 above)
Comment (dnhatn) [reply client/rest-high-level/src/main/java/org/elasticsearch/client/IndicesClient.java]: I pushed https://github.com/elastic/elasticsearch/pull/31161/commits/7117a3d90eb97f70c6ee7f64be430ec7d5d005ac

File: client/rest-high-level/src/main/java/org/elasticsearch/client/IndicesClient.java (hunk 3 above)
Comment (dnhatn) [reply client/rest-high-level/src/main/java/org/elasticsearch/client/IndicesClient.java]: I pushed https://github.com/elastic/elasticsearch/pull/31161/commits/7117a3d90eb97f70c6ee7f64be430ec7d5d005ac

File: server/src/test/java/org/elasticsearch/action/admin/indices/template/get/GetIndexTemplatesResponseTests.java (hunk 4 above)
Comment (javanna) [reply server/src/test/java/org/elasticsearch/action/admin/indices/template/get/GetIndexTemplatesResponseTests.java]: ok can you add a comment here then?

Comment (dnhatn) [reply server/src/test/java/org/elasticsearch/action/admin/indices/template/get/GetIndexTemplatesResponseTests.java]: Yes, I added a comment for this.

**Comment by dnhatn** [comment]
Thanks @cbuescher and @javanna for reviewing!

**Pull request merged**